io/zenoh-transport/src/unicast/transport.rs

```diff
@@ -393,28 +390,29 @@ impl TransportUnicastInner {
         let mut pipeline = zlinkget!(zread!(self.links), link)
             .map(|l| l.pipeline.clone())
             .ok_or_else(|| zerror!("Cannot close Link {:?}: not found", link))?;
-        if let Some(p) = pipeline.take() {
-            // Close message to be sent on the target link
-            let peer_id = Some(self.config.manager.zid());
-            let reason_id = reason;
-            let link_only = true; // This is should always be true when closing a link
-            let attachment = None; // No attachment here
-            let msg = TransportMessage::make_close(peer_id, reason_id, link_only, attachment);
+        // Close message to be sent on the target link
+        let peer_id = Some(self.config.manager.zid());
+        let reason_id = reason;
+        let link_only = true; // This is should always be true when closing a link
+        let attachment = None; // No attachment here
+        let msg = TransportMessage::make_close(peer_id, reason_id, link_only, attachment);
 
+        if let Some(p) = pipeline.take() {
             p.push_transport_message(msg, Priority::Background);
         }
+
         // Remove the link from the channel
         self.del_link(link).await
     }
 
     pub(crate) async fn close(&self, reason: u8) -> ZResult<()> {
         log::trace!("Closing transport with peer: {}", self.config.zid);
 
-        let mut pipelines: Vec<Arc<TransmissionPipeline>> = zread!(self.links)
+        let mut pipelines = zread!(self.links)
             .iter()
-            .filter_map(|sl| sl.pipeline.clone())
-            .collect();
-        for p in pipelines.drain(..) {
+            .map(|sl| sl.pipeline.clone())
```

Comment:
why not `filter_map` here rather than bellow? I guess it's likely they're all `Some` variants, but just in case some `None`s slip in there, might as well not risk extra allocations for them :)

Reply:
Fixed